File: code-orig/neural_xor.cpp
#include <iostream>
#include <chrono>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <iomanip>

using namespace std;

struct NeuralNetwork {
    vector<vector<double>> weights_input_hidden;
    vector<vector<double>> weights_hidden_output;
    size_t input_nodes, hidden_nodes, output_nodes;

    NeuralNetwork(size_t input_nodes, size_t hidden_nodes, size_t output_nodes)
        : input_nodes(input_nodes), hidden_nodes(hidden_nodes), output_nodes(output_nodes) {
        weights_input_hidden.resize(input_nodes, vector<double>(hidden_nodes));
        weights_hidden_output.resize(hidden_nodes, vector<double>(output_nodes));

        // Initialize weights with random values between -1 and 1
        srand(static_cast<unsigned>(time(nullptr)));
        for (size_t i = 0; i < input_nodes; ++i) {
            for (size_t j = 0; j < hidden_nodes; ++j) {
                weights_input_hidden[i][j] = static_cast<double>(rand()) / RAND_MAX * 2 - 1;
            }
        }
        for (size_t i = 0; i < hidden_nodes; ++i) {
            for (size_t j = 0; j < output_nodes; ++j) {
                weights_hidden_output[i][j] = static_cast<double>(rand()) / RAND_MAX * 2 - 1;
            }
        }
    }

    double sigmoid(double x) {
        return 1 / (1 + exp(-x));
    }

    double sigmoid_derivative(double x) {
        return x * (1 - x);
    }

    void train(vector<vector<double>>& inputs, vector<vector<double>>& targets, size_t epochs, double learning_rate) {
        ofstream loss_file("loss_data.txt");
        if (!loss_file.is_open()) {
            cerr << "Error opening loss_data.txt for writing" << endl;
            exit(EXIT_FAILURE);
        }

        for (size_t epoch = 0; epoch < epochs; ++epoch) {
            double total_loss = 0.0;
            for (size_t i = 0; i < inputs.size(); ++i) {
                vector<double> layer1_output(hidden_nodes, 0.0);
                vector<double> output(output_nodes, 0.0);

                // Forward pass
                for (size_t j = 0; j < hidden_nodes; ++j) {
                    for (size_t k = 0; k < input_nodes; ++k) {
                        layer1_output[j] += inputs[i][k] * weights_input_hidden[k][j];
                    }
                    layer1_output[j] = sigmoid(layer1_output[j]);
                }

                for (size_t j = 0; j < output_nodes; ++j) {
                    for (size_t k = 0; k < hidden_nodes; ++k) {
                        output[j] += layer1_output[k] * weights_hidden_output[k][j];
                    }
                    output[j] = sigmoid(output[j]);
                }

                vector<double> output_error(output_nodes, 0.0);
                vector<double> output_delta(output_nodes, 0.0);
                vector<double> hidden_error(hidden_nodes, 0.0);
                vector<double> hidden_delta(hidden_nodes, 0.0);

                // Backpropagation
                for (size_t j = 0; j < output_nodes; ++j) {
                    output_error[j] = targets[i][j] - output[j];
                    output_delta[j] = output_error[j] * sigmoid_derivative(output[j]);
                }

                for (size_t j = 0; j < hidden_nodes; ++j) {
                    for (size_t k = 0; k < output_nodes; ++k) {
                        hidden_error[j] += output_delta[k] * weights_hidden_output[j][k];
                    }
                    hidden_delta[j] = hidden_error[j] * sigmoid_derivative(layer1_output[j]);
                }

                // Weight adjustments
                for (size_t j = 0; j < input_nodes; ++j) {
                    for (size_t k = 0; k < hidden_nodes; ++k) {
                        weights_input_hidden[j][k] += inputs[i][j] * hidden_delta[k] * learning_rate;
                    }
                }

                for (size_t j = 0; j < hidden_nodes; ++j) {
                    for (size_t k = 0; k < output_nodes; ++k) {
                        weights_hidden_output[j][k] += layer1_output[j] * output_delta[k] * learning_rate;
                    }
                }

                // Calculate and store the loss for each iteration
                double loss = calculate_loss(inputs[i], targets[i]);
                total_loss += loss;
            }

            // Calculate and store the average loss for the epoch
            double avg_loss = total_loss / inputs.size();
            loss_file << avg_loss << endl;
        }

        loss_file.close();
    }

    double calculate_loss(vector<double>& input, vector<double>& target) {
        vector<double> predictions = predict(input);
        double loss = pow(target[0] - predictions[0], 2);
        return loss;
    }

    vector<double> predict(vector<double>& input) {
        vector<double> layer1_output(hidden_nodes, 0.0);
        vector<double> output(output_nodes, 0.0);

        // Forward pass
        for (size_t j = 0; j < hidden_nodes; ++j) {
            for (size_t k = 0; k < input_nodes; ++k) {
                layer1_output[j] += input[k] * weights_input_hidden[k][j];
            }
            layer1_output[j] = sigmoid(layer1_output[j]);
        }

        for (size_t j = 0; j < output_nodes; ++j) {
            for (size_t k = 0; k < hidden_nodes; ++k) {
                output[j] += layer1_output[k] * weights_hidden_output[k][j];
            }
            output[j] = sigmoid(output[j]);
        }

        return output;
    }

    void evaluatePerformance(vector<vector<double>>& testInputs, vector<vector<double>>& testTargets) {
        double correctPredictions = 0.0;

        for (size_t i = 0; i < testInputs.size(); ++i) {
            vector<double>& input = testInputs[i];
            vector<double>& target = testTargets[i];

            vector<double> prediction = predict(input);

            // Assuming a binary classification problem
            if ((prediction[0] > 0.5 && target[0] > 0.5) || (prediction[0] <= 0.5 && target[0] <= 0.5)) {
                correctPredictions += 1.0;
            }
        }

        double accuracy = correctPredictions / testInputs.size();
        cout << "Accuracy on Test Dataset: " << accuracy * 100.0 << "%\n";
    }

void plotPredictionsVsInputs(vector<vector<double>>& testInputs, vector<vector<double>>& testTargets) {
    ofstream predictions_file("predictions_data.txt");
    if (!predictions_file.is_open()) {
        cerr << "Error opening predictions_data.txt for writing" << endl;
        exit(EXIT_FAILURE);
    }

    for (size_t i = 0; i < testInputs.size(); ++i) {
        vector<double> input = testInputs[i];
        vector<double> target = testTargets[i];
        vector<double> prediction = predict(input);

        predictions_file << input[0] << " " << input[1] << " " << target[0] << " " << prediction[0] << endl;
    }

    predictions_file.close();

    // Plotting predictions vs inputs using gnuplot
    ofstream gnuplot_script("plot_script.gp");
    if (!gnuplot_script.is_open()) {
        cerr << "Error opening plot_script.gp for writing" << endl;
        exit(EXIT_FAILURE);
    }

    gnuplot_script << "plot 'predictions_data.txt' using 1:2 with points title 'Inputs', "
                   << "'' using 1:3 with points title 'Targets', "
                   << "'' using 1:4 with points title 'Predictions'";

    gnuplot_script.close();

    system("gnuplot -p plot_script.gp");
}
};

int main() {
    auto start_training = chrono::high_resolution_clock::now();

    size_t input_nodes = 2;
    size_t hidden_nodes = 4;
    size_t output_nodes = 1;

    vector<vector<double>> training_inputs = {
        {0.0, 0.0},
        {0.0, 1.0},
        {1.0, 0.0},
        {1.0, 1.0}
    };

    vector<vector<double>> training_targets = {
        {0.0},
        {1.0},
        {1.0},
        {0.0}
    };

    NeuralNetwork neural_network(input_nodes, hidden_nodes, output_nodes);
    neural_network.train(training_inputs, training_targets, 10000, 0.1);

    auto training_duration = chrono::high_resolution_clock::now() - start_training;
    vector<vector<double>> test_inputs = {
        {0.0, 0.0},
        {0.0, 1.0},
        {1.0, 0.0},
        {1.0, 1.0}
    };

    for (size_t i = 0; i < test_inputs.size(); ++i) {
        vector<double> predictions = neural_network.predict(test_inputs[i]);
        cout << "Input: {" << test_inputs[i][0] << ", " << test_inputs[i][1] << "} => Prediction: " << predictions[0] << endl;
    }

    neural_network.plotPredictionsVsInputs(test_inputs, training_targets);

    cout << "Training Time: " << chrono::duration_cast<chrono::milliseconds>(training_duration).count() << " ms\n";

    neural_network.evaluatePerformance(test_inputs, training_targets);

    // Graph analysis - Plot the training loss graph
    system("gnuplot -p -e \"plot 'loss_data.txt' with lines title 'Training Loss Over Epochs'\"");

    return 0;
}

File: code-orig/neural_xor.rs
extern crate gnuplot;
use rand::Rng;
use std::time::Instant;
use gnuplot::{Figure, Caption, Color};
//use crate::gnuplot::AxesCommon;
use gnuplot::Graph;
use gnuplot::AxesCommon;

struct NeuralNetwork {
    weights_input_hidden: Vec<Vec<f64>>,
    weights_hidden_output: Vec<Vec<f64>>,
}

impl NeuralNetwork {
    fn new(input_nodes: usize, hidden_nodes: usize, output_nodes: usize) -> Self {
        let mut rng = rand::thread_rng();
        let weights_input_hidden = (0..input_nodes)
            .map(|_| {
                (0..hidden_nodes)
                    .map(|_| rng.gen_range(-1.0..1.0))
                    .collect()
            })
            .collect();
        let weights_hidden_output = (0..hidden_nodes)
            .map(|_| {
                (0..output_nodes)
                    .map(|_| rng.gen_range(-1.0..1.0))
                    .collect()
            })
            .collect();

        NeuralNetwork {
            weights_input_hidden,
            weights_hidden_output,
        }
    }

    fn sigmoid(x: f64) -> f64 {
        1.0 / (1.0 + (-x).exp())
    }

    fn sigmoid_derivative(x: f64) -> f64 {
        x * (1.0 - x)
    }

    fn train_and_analyze(&mut self, inputs: &Vec<Vec<f64>>, targets: &Vec<Vec<f64>>, epochs: usize, learning_rate: f64) {
        let mut losses = Vec::new();

        for epoch in 0..epochs {
            for (input, target) in inputs.iter().zip(targets.iter()) {
                // Forward pass
                let mut layer1_output = vec![0.0; self.weights_input_hidden[0].len()];
                for i in 0..layer1_output.len() {
                    for j in 0..input.len() {
                        layer1_output[i] += input[j] * self.weights_input_hidden[j][i];
                    }
                    layer1_output[i] = NeuralNetwork::sigmoid(layer1_output[i]);
                }

                let mut output = vec![0.0; self.weights_hidden_output[0].len()];
                for i in 0..output.len() {
                    for j in 0..layer1_output.len() {
                        output[i] += layer1_output[j] * self.weights_hidden_output[j][i];
                    }
                }

                // Backpropagation
                let mut output_error = vec![0.0; target.len()];
                for i in 0..output_error.len() {
                    output_error[i] = target[i] - output[i];
                }

                let mut output_delta = vec![0.0; output.len()];
                for i in 0..output_delta.len() {
                    output_delta[i] = output_error[i] * NeuralNetwork::sigmoid_derivative(output[i]);
                }

                let mut hidden_error = vec![0.0; layer1_output.len()];
                for i in 0..hidden_error.len() {
                    for j in 0..output.len() {
                        hidden_error[i] += output_delta[j] * self.weights_hidden_output[i][j];
                    }
                }

                let mut hidden_delta = vec![0.0; layer1_output.len()];
                for i in 0..hidden_delta.len() {
                    hidden_delta[i] = hidden_error[i] * NeuralNetwork::sigmoid_derivative(layer1_output[i]);
                }

                // Update weights
                for i in 0..self.weights_input_hidden.len() {
                    for j in 0..self.weights_input_hidden[0].len() {
                        self.weights_input_hidden[i][j] += input[i] * hidden_delta[j] * learning_rate;
                    }
                }

                for i in 0..self.weights_hidden_output.len() {
                    for j in 0..self.weights_hidden_output[0].len() {
                        self.weights_hidden_output[i][j] += layer1_output[i] * output_delta[j] * learning_rate;
                    }
                }
            }

            // Calculate and store the loss for each epoch
            let loss = calculate_loss(self, inputs, targets);
            losses.push(loss);
        }

        // Plot the training loss graph
        plot_loss_graph(&losses);
    }

    fn predict(&self, inputs: &Vec<Vec<f64>>) -> Vec<Vec<f64>> {
        let mut predictions = vec![];
        for input in inputs {
            let mut layer1_output = vec![0.0; self.weights_input_hidden[0].len()];
            for i in 0..layer1_output.len() {
                for j in 0..input.len() {
                    layer1_output[i] += input[j] * self.weights_input_hidden[j][i];
                }
                layer1_output[i] = NeuralNetwork::sigmoid(layer1_output[i]);
            }

            let mut output = vec![0.0; self.weights_hidden_output[0].len()];
            for i in 0..output.len() {
                for j in 0..layer1_output.len() {
                    output[i] += layer1_output[j] * self.weights_hidden_output[j][i];
                }
            }
            predictions.push(output);
        }
        predictions
    }

    fn evaluate_performance(&self, test_inputs: &Vec<Vec<f64>>, test_targets: &Vec<Vec<f64>>) {
        let predictions = self.predict(test_inputs);
        let accuracy = calculate_accuracy(&predictions, test_targets);

        println!("Accuracy on Test Dataset: {:.2}%", accuracy * 100.0);
    }
}

fn calculate_loss(network: &NeuralNetwork, inputs: &Vec<Vec<f64>>, targets: &Vec<Vec<f64>>) -> f64 {
    let predictions = network.predict(inputs);
    let loss: f64 = predictions
        .iter()
        .zip(targets.iter())
        .map(|(pred, target)| (target[0] - pred[0]).powi(2))
        .sum();
    
    loss / predictions.len() as f64
}

fn calculate_accuracy(predictions: &Vec<Vec<f64>>, targets: &Vec<Vec<f64>>) -> f64 {
    let correct_predictions = predictions
        .iter()
        .zip(targets.iter())
        .filter(|&(pred, target)| (pred[0] > 0.5 && target[0] > 0.5) || (pred[0] <= 0.5 && target[0] <= 0.5))
        .count() as f64;

    correct_predictions / predictions.len() as f64
}

fn plot_predictions(predictions: &Vec<Vec<f64>>) {
    let mut fg = Figure::new();
    fg.set_terminal("pngcairo", "output.png"); // Specify output format and file name
    fg.axes2d()
        .set_title("predictions made by the neural network for each XOR input", &[])
        .set_legend(Graph(0.5), Graph(0.9), &[], &[])
        .set_x_label("training inputs(decimal)", &[])
        .set_y_label("predictions", &[])
        .lines(
            (0..4).map(|i| i as f64),
            predictions.iter().map(|p| p[0]),
            &[],
        );

    fg.show().unwrap();
}

fn plot_loss_graph(losses: &Vec<f64>) {
    let mut fg = Figure::new();
    fg.set_terminal("pngcairo", "loss_plot.png");

    fg.axes2d()
        .set_title("Training Loss Over Epochs", &[])
        .lines(0..losses.len() as u32, losses, &[Color("blue")]);

    fg.show();
}

fn main() {
    let start_training = Instant::now();
    let training_inputs = vec![
        vec![0.0, 0.0],
        vec![0.0, 1.0],
        vec![1.0, 0.0],
        vec![1.0, 1.0],
    ];
    let training_targets = vec![
        vec![0.0],
        vec![1.0],
        vec![1.0],
        vec![0.0],
    ];

    let mut neural_network = NeuralNetwork::new(2, 4, 1);
    neural_network.train_and_analyze(&training_inputs, &training_targets, 10000, 0.1);
    let training_duration = start_training.elapsed();

    let test_inputs = vec![
        vec![0.0, 0.0],
        vec![0.0, 1.0],
        vec![1.0, 0.0],
        vec![1.0, 1.0],
    ];

    let predictions = neural_network.predict(&test_inputs);

    for (input, prediction) in test_inputs.iter().zip(predictions.iter()) {
        println!(
            "Input: {{{}, {}}} => Prediction: {:.6}",
            input[0], input[1], prediction[0]
        );
    }

    println!("Training Time: {:?}", training_duration);

    neural_network.evaluate_performance(&test_inputs, &training_targets);
    plot_predictions(&predictions);
}